Mathlib/Analysis/AbsoluteValue/Equivalence.lean

```diff
@@ -151,6 +151,128 @@ theorem exists_one_lt_lt_one_of_not_isEquiv {v w : AbsoluteValue R S} (hv : v.Is
+  choose rₙ hrₙ using fun j hj ↦ hcⱼ j hj (.Iio 1) (by simpa using ⟨1, .inr rfl⟩) (by simp)
+  choose rN hrN using Filter.eventually_atTop.1 <| Filter.Tendsto.eventually_lt_const hbw hcₙ
+  let r := max (Finset.univ.sup fun j ↦ if h : j = i then r₁ else rₙ j h) rN
+  refine ⟨c r, hr₁ r ?_, fun j hj ↦ ?_, ?_⟩
+  · exact le_max_iff.2 <| .inl <|
+      Finset.le_sup_dite_pos (p := fun j ↦ j = i) (f := fun _ _ ↦ r₁) (Finset.mem_univ _) rfl
+  · exact hrₙ j hj _ <| le_max_iff.2 <| .inl <|
+      Finset.le_sup_dite_neg (fun j ↦ j = i) (Finset.mem_univ j) _
+  · exact hrN _ <| le_max_iff.2 (.inr le_rfl)
+
+open Fintype Subtype in
+/--
+If `v : ι → AbsoluteValue R S` is a finite collection of non-trivial and pairwise inequivalent
+absolute values, then for any `v i` there is some `a : R` such that `1 < v i a` while all other
+`v j a < 1`.
+-/
+theorem exists_one_lt_lt_one_pi_of_not_isEquiv (h : ∀ i, (v i).IsNontrivial)
+    (hv : Pairwise fun i j ↦ ¬(v i).IsEquiv (v j)) :
+    ∀ i, ∃ (a : R), 1 < v i a ∧ ∀ j ≠ i, v j a < 1 := by
+  let P (ι : Type u_3) [Fintype ι] : Prop := [DecidableEq ι] →
+    ∀ v : ι → AbsoluteValue R S, (∀ i, (v i).IsNontrivial) →
+      (Pairwise fun i j ↦ ¬(v i).IsEquiv (v j)) → ∀ i, ∃ (a : R), 1 < v i a ∧ ∀ j ≠ i, v j a < 1
+  -- Use strong induction on the index.
+  revert hv h; refine induction_subsingleton_or_nontrivial (P := P) ι (fun ι _ _ _ v h hv i ↦ ?_)
+    (fun ι _ _ ih _ v h hv i ↦ ?_) v
+  · -- If `ι` is trivial this follows immediately from `(v i).IsNontrivial`.
+    let ⟨a, ha⟩ := (h i).exists_abv_gt_one
+    exact ⟨a, ha, fun j hij ↦ absurd (Subsingleton.elim i j) hij.symm⟩
+  · rcases eq_or_ne (card ι) 2 with (hc | hc)
+    · -- If `ι` has two elements this is `exists_one_lt_lt_one_of_not_isEquiv`.
+      let ⟨j, hj⟩ := (Nat.card_eq_two_iff' i).1 <| card_eq_nat_card ▸ hc
+      let ⟨a, ha⟩ := (v i).exists_one_lt_lt_one_of_not_isEquiv (h i) (h j) (hv hj.1.symm)
+      exact ⟨a, ha.1, fun _ h ↦ hj.2 _ h ▸ ha.2⟩
+    have hlt : 2 < card ι := Nat.lt_of_le_of_ne (one_lt_card_iff_nontrivial.2 ‹_›) hc.symm
+    -- Otherwise, choose another distinguished index `j ≠ i`.
+    let ⟨j, hj⟩ := exists_ne i
+    -- Apply induction first on the subcollection `v i` for `i ≠ j` to get `a : K`
+    let ⟨a, ha⟩ := ih {k : ι // k ≠ j} (card_subtype_lt fun a ↦ a rfl) (restrict _ v)
+      (fun i ↦ h _) (hv.comp_of_injective val_injective) ⟨i, hj.symm⟩
+    -- Then apply induction next to the subcollection `{v i, v j}` to get `b : K`.
+    let ⟨b, hb⟩ := ih {k : ι // k = i ∨ k = j} (by linarith [card_subtype_or_eq hj.symm])
+      (restrict _ v) (fun _ ↦ h _) (hv.comp_of_injective val_injective) ⟨i, .inl rfl⟩
+    rcases eq_or_ne (v j a) 1 with (ha₁ | ha₁)
+    · -- If `v j a = 1` then take a large enough value from the sequence `a ^ n * b`.
+      let ⟨c, hc⟩ := exists_one_lt_lt_one_pi_of_eq_one ha.1 ha.2 ha₁ hb.1 (hb.2 ⟨j, .inr rfl⟩
+        (coe_ne_coe.1 hj))
```

Comment:
```suggestion
        (by grind)
```
I suspect a couple of trivial things can be golfed using `grind`. It seems a good idea.